lib/Conversion/LWEToPolynomial/LWEToPolynomial.cpp

```diff
@@ -0,0 +1,126 @@
+#include "mlir/include/mlir/Support/LLVM.h"              // from @llvm-project
+#include "mlir/include/mlir/Support/LogicalResult.h"     // from @llvm-project
+#include "mlir/include/mlir/Transforms/DialectConversion.h"  // from @llvm-project
+
+namespace mlir::heir::lwe {
+
+#define GEN_PASS_DEF_LWETOPOLYNOMIAL
+#include "lib/Conversion/LWEToPolynomial/LWEToPolynomial.h.inc"
+
+class CiphertextTypeConverter : public TypeConverter {
+ public:
+  // Convert ciphertext to tensor<#dim x !poly.poly<#rings[#level]>>
+  // Our precondition (not verified) is that the key and plaintext have only one
+  // dimension, and the ciphertext has two dimensions. (i.e.) The key is a
+  // single polynomial and not a tensor of higher dimensions. We may support
+  // higher dimensions in the future (for schemes such as TFHE). Here are some
+  // TODO:
+  // 1. Add an encrypt/decrypt verifier to check for the preconditions of
+  // encrypt and decrypt. See lib/Target/Verilog/VerilgEmitter.cpp for an
+  // example of op->walk to use in verifying.
+  // 2. For TFHE, which can support higher dimensional keys, plaintexts, and
+  // ciphertexts, we need to add support for encrypt and decrypt for
+  // those cases.
+  CiphertextTypeConverter(MLIRContext *ctx) {
+    addConversion([](Type type) { return type; });
+    addConversion([ctx](lwe::RLWECiphertextType type) -> Type {
+      auto rlweParams = type.getRlweParams();
+      auto ring = rlweParams.getRing();
+      auto polyTy = ::mlir::polynomial::PolynomialType::get(ctx, ring);
+
+      return RankedTensorType::get({rlweParams.getDimension()}, polyTy);
+    });
+    addConversion([ctx](lwe::RLWEPlaintextType type) -> Type {
+      auto ring = type.getRing();
+      auto polyTy = ::mlir::polynomial::PolynomialType::get(ctx, ring);
+      return polyTy;
+    });
+    addConversion([ctx](lwe::RLWESecretKeyType type) -> Type {
+      auto ring = type.getRlweParams().getRing();
+      auto polyTy = ::mlir::polynomial::PolynomialType::get(ctx, ring);
+      return polyTy;
+    });
+    addConversion([ctx](lwe::RLWEPublicKeyType type) -> Type {
+      auto rlweParams = type.getRlweParams();
+      auto ring = type.getRlweParams().getRing();
+      auto polyTy = ::mlir::polynomial::PolynomialType::get(ctx, ring);
+
+      return RankedTensorType::get({rlweParams.getDimension()}, polyTy);
+    });
+  }
+};
+
+struct ConvertRLWEDecrypt : public OpConversionPattern<RLWEDecryptOp> {
+  ConvertRLWEDecrypt(mlir::MLIRContext *context)
+      : OpConversionPattern<RLWEDecryptOp>(context) {}
+
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult matchAndRewrite(
+      RLWEDecryptOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+    auto loc = op.getLoc();
+    auto input = adaptor.getInput();
+    auto secretKey = adaptor.getSecretKey();
+
+    ImplicitLocOpBuilder builder(loc, rewriter);
+
+    auto index0 = builder.create<arith::ConstantIndexOp>(0);
```

Comment:
nit: add a quick docstring style comment on the computation performed like
```
// For a ciphertext input = (c_0, c_1), calculates
// plaintext = secretKey * c_0 + c_1
```

Reply:
Ack.